MetaMask/metamask-extension
# ZK Deterministic Signature RPC method (Plume)

## Explanation



Adds new `eth_getPlumeSignature` RPC method that implements a [novel ECDSA nullifier scheme](https://aayushg.com/thesis.pdf).

The `eth_getPlumeSignature` method takes in two parameters, a **message** and an **address**, then generates a deterministic signature (plume) and several other inputs. The plume can be used as a nullifier to prevent double-spending in an anonymity set. This capability unlocks novel on-chain behavior, such as [private DAO voting](https://prop.house/nouns/private-voting-research-sprint), [fair, non-doxxing airdrops](https://github.com/stealthdrop/stealthdrop), and more.

## Screenshots/Screencaps


https://user-images.githubusercontent.com/36896271/224298214-d12aed6a-4445-4deb-9093-6bc85b5893ab.mov



### Before

RPC method did not exist!


### After



<img width="364" alt="plume-confirmation-window" src="https://user-images.githubusercontent.com/36896271/224298304-d3c34a04-40fd-4db9-b0ff-b5732388a3ce.png">

## Manual Testing Steps



After building and running MetaMask locally, enter this into the browser console
```javascript
window.ethereum.request({
    method: 'eth_getPlumeSignature',
    params: [
        'this is a message to sign',
        '<YOUR ETHEREUM ADDRESS>'
    ]
})
    .then(data => console.log(data))
```

A confirmation screen should open up. After clicking confirm, you will see the plume and other signals outputted into the console.

See screen capture above for more details.

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
